Fails to build with Rust stable

Hello! 👋🏼 

I [packaged](https://archlinux.org/packages/extra/x86_64/hawkeye/) this tool for Arch Linux, however I'm hitting:

```
error[E0554]: `#![feature]` may not be used on the stable release channel
  --> fmt/src/lib.rs:15:1
   |
15 | #![feature(extract_if)]
   | ^^^^^^^^^^^^^^^^^^^^^^^

error[E0554]: `#![feature]` may not be used on the stable release channel
  --> fmt/src/lib.rs:16:1
   |
16 | #![feature(let_chains)]
   | ^^^^^^^^^^^^^^^^^^^^^^^

error[E0554]: `#![feature]` may not be used on the stable release channel
  --> fmt/src/lib.rs:15:12
   |
15 | #![feature(extract_if)]
   |            ^^^^^^^^^^
```

I solved this by using `cargo-nightly` package but this unfortunately effects reproducibility :/

I was wondering if it is possible to support Rust stable 👀

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -40,6 +40,7 @@ jobs:
       - uses: actions/checkout@v4
       - uses: korandoru/hawkeye@v4
       - uses: Swatinem/rust-cache@v2
+      - uses: dtolnay/rust-toolchain@nightly
       - name: Check Clippy
         run: cargo clippy --tests --all-features --all-targets --workspace -- -D warnings
       - name: Check format

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -26,6 +26,7 @@ authors = ["tison <wander4096@gmail.com>"]
 readme = "README.md"
 license = "Apache-2.0"
 repository = "https://github.com/korandoru/hawkeye/"
+rust-version = "1.76.0"
 
 [workspace.dependencies]
 hawkeye-fmt = { version = "5.2.1", path = "fmt" }

--- a/cli/src/main.rs
+++ b/cli/src/main.rs
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#![feature(extract_if)]
+
 use clap::{crate_description, FromArgMatches, Subcommand};
 use tracing::level_filters::LevelFilter;
 use tracing_subscriber::{fmt, layer::SubscriberExt, util::SubscriberInitExt, EnvFilter};

--- a/cli/src/main.rs
+++ b/cli/src/main.rs
@@ -12,8 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#![feature(extract_if)]
-
 use clap::{crate_description, FromArgMatches, Subcommand};
 use tracing::level_filters::LevelFilter;
 use tracing_subscriber::{fmt, layer::SubscriberExt, util::SubscriberInitExt, EnvFilter};

--- a/fmt/src/document/mod.rs
+++ b/fmt/src/document/mod.rs
@@ -42,7 +42,7 @@ impl Document {
         keywords: &[String],
         properties: HashMap<String, String>,
     ) -> std::io::Result<Self> {
-        let parser = header::parser::parse_header(&filepath, header_def.clone(), keywords)?;
+        let parser = header::parser::parse_header(&filepath, &header_def, keywords)?;
         Ok(Self {
             filepath,
             header_def,

--- a/fmt/src/header/parser.rs
+++ b/fmt/src/header/parser.rs
@@ -43,20 +43,18 @@ pub fn parse_header(
         let mut begin_pos = 0;
         if header_def.skip_line_pattern.is_some() {
             // the format expect to find lines to be skipped
-            while let Some(l) = line.as_ref() {
-                if header_def.is_skip_line(l) {
-                    break;
-                }
+            while let Some(l) = line.as_ref()
+                && !header_def.is_skip_line(l)
+            {
                 begin_pos = file_content.pos;
                 line = file_content.next_line();
             }
 
             // at least we have found the line to skip, or we are the end of the file
             // this time we are going to skip next lines if they match the skip pattern
-            while let Some(l) = line.as_ref() {
-                if !header_def.is_skip_line(l) {
-                    break;
-                }
+            while let Some(l) = line.as_ref()
+                && header_def.is_skip_line(l)
+            {
                 begin_pos = file_content.pos;
                 line = file_content.next_line();
             }
@@ -73,22 +71,19 @@ pub fn parse_header(
     };
 
     // 2. has header
-    let mut existing_header = || -> bool {
+    let existing_header = {
         // skip blank lines
-        while let Some(l) = line.as_ref() {
-            if !l.trim().is_empty() {
-                break;
-            }
+        while let Some(l) = line.as_ref()
+            && l.trim().is_empty()
+        {
             line = file_content.next_line();
         }
 
         // check if there is already a header
         let mut got_header = false;
-        if let Some(l) = line.as_ref() {
-            if header_def.is_first_header_line(l) {
-                return false;
-            }
-
+        if let Some(l) = line.as_ref()
+            && header_def.is_first_header_line(l)
+        {
             let mut in_place_header = String::new();
             in_place_header.push_str(&l.to_lowercase());
 
@@ -96,10 +91,9 @@ pub fn parse_header(
 
             // skip blank lines before header text
             if header_def.allow_blank_lines {
-                while let Some(l) = line.as_ref() {
-                    if !l.trim().is_empty() {
-                        break;
-                    }
+                while let Some(l) = line.as_ref()
+                    && l.trim().is_empty()
+                {
                     line = file_content.next_line();
                 }
             }
@@ -126,15 +120,13 @@ pub fn parse_header(
                     } else {
                         loop {
                             line = file_content.next_line();
-                            if let Some(l) = line.as_ref() {
-                                if l.starts_with(before) {
-                                    in_place_header.push_str(&l.to_lowercase());
-                                    if header_def.multiple_lines
-                                        && header_def.is_last_header_line(l)
-                                    {
-                                        found_end = true;
-                                        break;
-                                    }
+                            if let Some(l) = line.as_ref()
+                                && l.starts_with(before)
+                            {
+                                in_place_header.push_str(&l.to_lowercase());
+                                if header_def.multiple_lines && header_def.is_last_header_line(l) {
+                                    found_end = true;
+                                    break;
                                 }
                             } else {
                                 break;
@@ -166,13 +158,11 @@ pub fn parse_header(
                     // before each line
                     let pos = file_content.pos;
                     // check if the line is the end line
-                    while let Some(l) = line.as_ref() {
-                        if header_def.is_last_header_line(l)
-                            || (!header_def.allow_blank_lines && l.trim().is_empty())
-                            || !l.starts_with(before)
-                        {
-                            break;
-                        }
+                    while let Some(l) = line.as_ref()
+                        && !header_def.is_last_header_line(l)
+                        && (header_def.allow_blank_lines || !l.trim().is_empty())
+                        && l.starts_with(before)
+                    {
                         line = file_content.next_line();
                     }
                     if line.is_none() {
@@ -208,15 +198,14 @@ pub fn parse_header(
     };
 
     // 3. find end position
-    let end_pos = if existing_header() {
+    let end_pos = if existing_header {
         // we check if there is a header, if the next line is the blank line of the header
         let mut end = file_content.pos;
         line = file_content.next_line();
         if begin_pos == 0 {
-            while let Some(l) = line.as_ref() {
-                if !l.trim().is_empty() {
-                    break;
-                }
+            while let Some(l) = line.as_ref()
+                && l.trim().is_empty()
+            {
                 end = file_content.pos;
                 line = file_content.next_line();
             }
@@ -262,8 +251,9 @@ impl FileContent {
                 let mut content = String::new();
                 let mut reader = File::open(file).map(BufReader::new)?;
                 let mut buf = String::new();
-                let mut n = reader.read_line(&mut buf)?;
-                while n > 0 {
+                while let n = reader.read_line(&mut buf)?
+                    && n > 0
+                {
                     if buf.ends_with('\n') {
                         buf.pop();
                         if buf.ends_with('\r') {
@@ -275,7 +265,6 @@ impl FileContent {
                         content.push_str(&buf);
                     }
                     buf.clear();
-                    n = reader.read_line(&mut buf)?;
                 }
                 content
             },

--- a/fmt/src/header/parser.rs
+++ b/fmt/src/header/parser.rs
@@ -26,176 +26,22 @@ pub struct HeaderParser {
     pub begin_pos: usize,
     /// Some if header exists; None if header does not exist.
     pub end_pos: Option<usize>,
-    pub header_def: HeaderDef,
     pub file_content: FileContent,
 }
 
 pub fn parse_header(
     file: &Path,
-    header_def: HeaderDef,
+    header_def: &HeaderDef,
     keywords: &[String],
 ) -> std::io::Result<HeaderParser> {
     let mut file_content = FileContent::new(file)?;
     let mut line = file_content.next_line();
 
     // 1. find begin position
-    let begin_pos = {
-        let mut begin_pos = 0;
-        if header_def.skip_line_pattern.is_some() {
-            // the format expect to find lines to be skipped
-            while let Some(l) = line.as_ref()
-                && !header_def.is_skip_line(l)
-            {
-                begin_pos = file_content.pos;
-                line = file_content.next_line();
-            }
-
-            // at least we have found the line to skip, or we are the end of the file
-            // this time we are going to skip next lines if they match the skip pattern
-            while let Some(l) = line.as_ref()
-                && header_def.is_skip_line(l)
-            {
-                begin_pos = file_content.pos;
-                line = file_content.next_line();
-            }
-
-            // After skipping everything we are at the end of the file
-            // Header has to be at the file beginning
-            if line.is_none() {
-                begin_pos = 0;
-                file_content.reset();
-                line = file_content.next_line();
-            }
-        }
-        begin_pos
-    };
+    let begin_pos = find_first_position(&mut line, &mut file_content, header_def);
 
     // 2. has header
-    let existing_header = {
-        // skip blank lines
-        while let Some(l) = line.as_ref()
-            && l.trim().is_empty()
-        {
-            line = file_content.next_line();
-        }
-
-        // check if there is already a header
-        let mut got_header = false;
-        if let Some(l) = line.as_ref()
-            && header_def.is_first_header_line(l)
-        {
-            let mut in_place_header = String::new();
-            in_place_header.push_str(&l.to_lowercase());
-
-            line = file_content.next_line();
-
-            // skip blank lines before header text
-            if header_def.allow_blank_lines {
-                while let Some(l) = line.as_ref()
-                    && l.trim().is_empty()
-                {
-                    line = file_content.next_line();
-                }
-            }
-
-            // first header detected line & potential blank lines have been detected
-            // following lines should be header lines
-            if let Some(l) = line.as_ref() {
-                in_place_header.push_str(&l.to_lowercase());
-
-                let before = {
-                    let mut before = header_def.before_each_line.trim_end();
-                    if before.is_empty() && !header_def.multiple_lines {
-                        before = header_def.before_each_line.as_str();
-                    }
-                    before
-                };
-
-                let found_end = {
-                    let mut found_end = false;
-                    if (header_def.multiple_lines && header_def.is_last_header_line(l))
-                        || l.trim().is_empty()
-                    {
-                        found_end = true;
-                    } else {
-                        loop {
-                            line = file_content.next_line();
-                            if let Some(l) = line.as_ref()
-                                && l.starts_with(before)
-                            {
-                                in_place_header.push_str(&l.to_lowercase());
-                                if header_def.multiple_lines && header_def.is_last_header_line(l) {
-                                    found_end = true;
-                                    break;
-                                }
-                            } else {
-                                break;
-                            }
-                        }
-
-                        if line.as_ref().map(|l| l.trim().is_empty()).unwrap_or(true) {
-                            found_end = true;
-                        }
-                    }
-                    found_end
-                };
-
-                // skip blank lines after header text
-                if header_def.multiple_lines && header_def.allow_blank_lines && !found_end {
-                    loop {
-                        line = file_content.next_line();
-                        if !line.as_ref().map(|l| l.trim().is_empty()).unwrap_or(false) {
-                            break;
-                        }
-                    }
-                    file_content.rewind();
-                } else if !header_def.multiple_lines && !found_end {
-                    file_content.rewind();
-                }
-
-                if !header_def.multiple_lines {
-                    // keep track of the position for headers where the end line is the same as the
-                    // before each line
-                    let pos = file_content.pos;
-                    // check if the line is the end line
-                    while let Some(l) = line.as_ref()
-                        && !header_def.is_last_header_line(l)
-                        && (header_def.allow_blank_lines || !l.trim().is_empty())
-                        && l.starts_with(before)
-                    {
-                        line = file_content.next_line();
-                    }
-                    if line.is_none() {
-                        file_content.reset_to(pos);
-                    }
-                } else if line.is_some() {
-                    // we could end up there if we still have some lines, but not matching "before".
-                    // This can be the last line in a multi line header
-                    let pos = file_content.pos;
-                    line = file_content.next_line();
-                    if line
-                        .as_ref()
-                        .map(|l| !header_def.is_last_header_line(l))
-                        .unwrap_or(true)
-                    {
-                        file_content.reset_to(pos);
-                    }
-                }
-
-                got_header = true;
-                for keyword in keywords {
-                    if !in_place_header.contains(keyword) {
-                        got_header = false;
-                        break;
-                    }
-                }
-            }
-            // else - we detected previously a one line comment block that matches the header
-            // detection it is not a header it is a comment
-        }
-
-        got_header
-    };
+    let existing_header = existing_header(&mut line, &mut file_content, header_def, keywords);
 
     // 3. find end position
     let end_pos = if existing_header {
@@ -203,9 +49,7 @@ pub fn parse_header(
         let mut end = file_content.pos;
         line = file_content.next_line();
         if begin_pos == 0 {
-            while let Some(l) = line.as_ref()
-                && l.trim().is_empty()
-            {
+            while line.as_ref().map(|l| l.trim().is_empty()).unwrap_or(false) {
                 end = file_content.pos;
                 line = file_content.next_line();
             }
@@ -223,9 +67,177 @@ pub fn parse_header(
     Ok(HeaderParser {
         begin_pos,
         end_pos,
-        header_def,
         file_content,
     })
+}
+
+fn find_first_position(
+    line: &mut Option<String>,
+    file_content: &mut FileContent,
+    header_def: &HeaderDef,
+) -> usize {
+    let mut begin_pos = 0;
+    if header_def.skip_line_pattern.is_some() {
+        // the format expect to find lines to be skipped
+        while line
+            .as_ref()
+            .map(|l| !header_def.is_skip_line(l))
+            .unwrap_or(false)
+        {
+            begin_pos = file_content.pos;
+            *line = file_content.next_line();
+        }
+
+        // at least we have found the line to skip, or we are the end of the file
+        // this time we are going to skip next lines if they match the skip pattern
+        while line
+            .as_ref()
+            .map(|l| header_def.is_skip_line(l))
+            .unwrap_or(false)
+        {
+            begin_pos = file_content.pos;
+            *line = file_content.next_line();
+        }
+
+        // After skipping everything we are at the end of the file
+        // Header has to be at the file beginning
+        if line.is_none() {
+            begin_pos = 0;
+            file_content.reset();
+            *line = file_content.next_line();
+        }
+    }
+    begin_pos
+}
+
+fn existing_header(
+    line: &mut Option<String>,
+    file_content: &mut FileContent,
+    header_def: &HeaderDef,
+    keywords: &[String],
+) -> bool {
+    // skip blank lines
+    while line.as_ref().map(|l| l.trim().is_empty()).unwrap_or(false) {
+        *line = file_content.next_line();
+    }
+
+    // check if there is already a header
+    let l = match line.as_ref() {
+        Some(l) if header_def.is_first_header_line(l) => l,
+        _ => return false,
+    };
+
+    let mut got_header = false;
+    let mut in_place_header = String::new();
+    in_place_header.push_str(&l.to_lowercase());
+
+    *line = file_content.next_line();
+
+    // skip blank lines before header text
+    if header_def.allow_blank_lines {
+        while line.as_ref().map(|l| l.trim().is_empty()).unwrap_or(false) {
+            *line = file_content.next_line();
+        }
+    }
+
+    // first header detected line & potential blank lines have been detected
+    // following lines should be header lines
+    if let Some(l) = line.as_ref() {
+        in_place_header.push_str(&l.to_lowercase());
+
+        let before = {
+            let mut before = header_def.before_each_line.trim_end();
+            if before.is_empty() && !header_def.multiple_lines {
+                before = header_def.before_each_line.as_str();
+            }
+            before
+        };
+
+        let found_end = {
+            let mut found_end = false;
+            if (header_def.multiple_lines && header_def.is_last_header_line(l))
+                || l.trim().is_empty()
+            {
+                found_end = true;
+            } else {
+                loop {
+                    *line = file_content.next_line();
+                    match line.as_ref() {
+                        Some(l) if l.starts_with(before) => {
+                            in_place_header.push_str(&l.to_lowercase());
+                            if header_def.multiple_lines && header_def.is_last_header_line(l) {
+                                found_end = true;
+                                break;
+                            }
+                        }
+                        _ => break,
+                    }
+                }
+
+                if line.as_ref().map(|l| l.trim().is_empty()).unwrap_or(true) {
+                    found_end = true;
+                }
+            }
+            found_end
+        };
+
+        // skip blank lines after header text
+        if header_def.multiple_lines && header_def.allow_blank_lines && !found_end {
+            loop {
+                *line = file_content.next_line();
+                if !line.as_ref().map(|l| l.trim().is_empty()).unwrap_or(false) {
+                    break;
+                }
+            }
+            file_content.rewind();
+        } else if !header_def.multiple_lines && !found_end {
+            file_content.rewind();
+        }
+
+        if !header_def.multiple_lines {
+            // keep track of the position for headers where the end line is the same as the
+            // before each line
+            let pos = file_content.pos;
+            // check if the line is the end line
+            while line
+                .as_ref()
+                .map(|l| {
+                    !header_def.is_last_header_line(l)
+                        && (header_def.allow_blank_lines || !l.trim().is_empty())
+                        && l.starts_with(before)
+                })
+                .unwrap_or(false)
+            {
+                *line = file_content.next_line();
+            }
+            if line.is_none() {
+                file_content.reset_to(pos);
+            }
+        } else if line.is_some() {
+            // we could end up there if we still have some lines, but not matching "before".
+            // This can be the last line in a multi line header
+            let pos = file_content.pos;
+            *line = file_content.next_line();
+            if line
+                .as_ref()
+                .map(|l| !header_def.is_last_header_line(l))
+                .unwrap_or(true)
+            {
+                file_content.reset_to(pos);
+            }
+        }
+
+        got_header = true;
+        for keyword in keywords {
+            if !in_place_header.contains(keyword) {
+                got_header = false;
+                break;
+            }
+        }
+    }
+    // else - we detected previously a one line comment block that matches the header
+    // detection it is not a header it is a comment
+    got_header
 }
 
 #[derive(Debug)]
@@ -251,9 +263,8 @@ impl FileContent {
                 let mut content = String::new();
                 let mut reader = File::open(file).map(BufReader::new)?;
                 let mut buf = String::new();
-                while let n = reader.read_line(&mut buf)?
-                    && n > 0
-                {
+                let mut n = reader.read_line(&mut buf)?;
+                while n > 0 {
                     if buf.ends_with('\n') {
                         buf.pop();
                         if buf.ends_with('\r') {
@@ -265,6 +276,7 @@ impl FileContent {
                         content.push_str(&buf);
                     }
                     buf.clear();
+                    n = reader.read_line(&mut buf)?;
                 }
                 content
             },

--- a/fmt/src/header/parser.rs
+++ b/fmt/src/header/parser.rs
@@ -43,18 +43,20 @@ pub fn parse_header(
         let mut begin_pos = 0;
         if header_def.skip_line_pattern.is_some() {
             // the format expect to find lines to be skipped
-            while let Some(l) = line.as_ref()
-                && !header_def.is_skip_line(l)
-            {
+            while let Some(l) = line.as_ref() {
+                if header_def.is_skip_line(l) {
+                    break;
+                }
                 begin_pos = file_content.pos;
                 line = file_content.next_line();
             }
 
             // at least we have found the line to skip, or we are the end of the file
             // this time we are going to skip next lines if they match the skip pattern
-            while let Some(l) = line.as_ref()
-                && header_def.is_skip_line(l)
-            {
+            while let Some(l) = line.as_ref() {
+                if !header_def.is_skip_line(l) {
+                    break;
+                }
                 begin_pos = file_content.pos;
                 line = file_content.next_line();
             }
@@ -71,19 +73,22 @@ pub fn parse_header(
     };
 
     // 2. has header
-    let existing_header = {
+    let mut existing_header = || -> bool {
         // skip blank lines
-        while let Some(l) = line.as_ref()
-            && l.trim().is_empty()
-        {
+        while let Some(l) = line.as_ref() {
+            if !l.trim().is_empty() {
+                break;
+            }
             line = file_content.next_line();
         }
 
         // check if there is already a header
         let mut got_header = false;
-        if let Some(l) = line.as_ref()
-            && header_def.is_first_header_line(l)
-        {
+        if let Some(l) = line.as_ref() {
+            if header_def.is_first_header_line(l) {
+                return false;
+            }
+
             let mut in_place_header = String::new();
             in_place_header.push_str(&l.to_lowercase());
 
@@ -91,9 +96,10 @@ pub fn parse_header(
 
             // skip blank lines before header text
             if header_def.allow_blank_lines {
-                while let Some(l) = line.as_ref()
-                    && l.trim().is_empty()
-                {
+                while let Some(l) = line.as_ref() {
+                    if !l.trim().is_empty() {
+                        break;
+                    }
                     line = file_content.next_line();
                 }
             }
@@ -120,13 +126,15 @@ pub fn parse_header(
                     } else {
                         loop {
                             line = file_content.next_line();
-                            if let Some(l) = line.as_ref()
-                                && l.starts_with(before)
-                            {
-                                in_place_header.push_str(&l.to_lowercase());
-                                if header_def.multiple_lines && header_def.is_last_header_line(l) {
-                                    found_end = true;
-                                    break;
+                            if let Some(l) = line.as_ref() {
+                                if l.starts_with(before) {
+                                    in_place_header.push_str(&l.to_lowercase());
+                                    if header_def.multiple_lines
+                                        && header_def.is_last_header_line(l)
+                                    {
+                                        found_end = true;
+                                        break;
+                                    }
                                 }
                             } else {
                                 break;
@@ -158,11 +166,13 @@ pub fn parse_header(
                     // before each line
                     let pos = file_content.pos;
                     // check if the line is the end line
-                    while let Some(l) = line.as_ref()
-                        && !header_def.is_last_header_line(l)
-                        && (header_def.allow_blank_lines || !l.trim().is_empty())
-                        && l.starts_with(before)
-                    {
+                    while let Some(l) = line.as_ref() {
+                        if header_def.is_last_header_line(l)
+                            || (!header_def.allow_blank_lines && l.trim().is_empty())
+                            || !l.starts_with(before)
+                        {
+                            break;
+                        }
                         line = file_content.next_line();
                     }
                     if line.is_none() {
@@ -198,14 +208,15 @@ pub fn parse_header(
     };
 
     // 3. find end position
-    let end_pos = if existing_header {
+    let end_pos = if existing_header() {
         // we check if there is a header, if the next line is the blank line of the header
         let mut end = file_content.pos;
         line = file_content.next_line();
         if begin_pos == 0 {
-            while let Some(l) = line.as_ref()
-                && l.trim().is_empty()
-            {
+            while let Some(l) = line.as_ref() {
+                if !l.trim().is_empty() {
+                    break;
+                }
                 end = file_content.pos;
                 line = file_content.next_line();
             }
@@ -251,9 +262,8 @@ impl FileContent {
                 let mut content = String::new();
                 let mut reader = File::open(file).map(BufReader::new)?;
                 let mut buf = String::new();
-                while let n = reader.read_line(&mut buf)?
-                    && n > 0
-                {
+                let mut n = reader.read_line(&mut buf)?;
+                while n > 0 {
                     if buf.ends_with('\n') {
                         buf.pop();
                         if buf.ends_with('\r') {
@@ -265,6 +275,7 @@ impl FileContent {
                         content.push_str(&buf);
                     }
                     buf.clear();
+                    n = reader.read_line(&mut buf)?;
                 }
                 content
             },

--- a/fmt/src/lib.rs
+++ b/fmt/src/lib.rs
@@ -12,8 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#![feature(let_chains)]
-
 use crate::error::Error;
 
 pub mod config;

--- a/fmt/src/lib.rs
+++ b/fmt/src/lib.rs
@@ -12,6 +12,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#![feature(extract_if)]
+#![feature(let_chains)]
+
 use crate::error::Error;
 
 pub mod config;

--- a/fmt/src/lib.rs
+++ b/fmt/src/lib.rs
@@ -12,9 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#![feature(extract_if)]
-#![feature(let_chains)]
-
 use crate::error::Error;
 
 pub mod config;

--- a/fmt/src/lib.rs
+++ b/fmt/src/lib.rs
@@ -12,7 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#![feature(extract_if)]
 #![feature(let_chains)]
 
 use crate::error::Error;

--- a/fmt/src/selection.rs
+++ b/fmt/src/selection.rs
@@ -75,14 +75,13 @@ impl Selection {
 
         let (excludes, reverse_excludes) = {
             let mut excludes = self.excludes;
-            let mut reverse_excludes = Vec::new();
-            excludes.retain(|pat| {
-                let keep_pat = pat.starts_with('!');
-                if !keep_pat {
-                    reverse_excludes.push(pat[1..].to_string());
-                }
-                keep_pat
-            });
+            let reverse_excludes = excludes
+                .extract_if(|pat| pat.starts_with('!'))
+                .map(|mut pat| {
+                    pat.remove(0);
+                    pat
+                })
+                .collect::<Vec<_>>();
             (excludes, reverse_excludes)
         };
 

--- a/fmt/src/selection.rs
+++ b/fmt/src/selection.rs
@@ -75,13 +75,14 @@ impl Selection {
 
         let (excludes, reverse_excludes) = {
             let mut excludes = self.excludes;
-            let reverse_excludes = excludes
-                .extract_if(|pat| pat.starts_with('!'))
-                .map(|mut pat| {
-                    pat.remove(0);
-                    pat
-                })
-                .collect::<Vec<_>>();
+            let mut reverse_excludes = Vec::new();
+            excludes.retain(|pat| {
+                let keep_pat = pat.starts_with('!');
+                if !keep_pat {
+                    reverse_excludes.push(pat[1..].to_string());
+                }
+                keep_pat
+            });
             (excludes, reverse_excludes)
         };
 

--- a/fmt/src/selection.rs
+++ b/fmt/src/selection.rs
@@ -75,13 +75,17 @@ impl Selection {
 
         let (excludes, reverse_excludes) = {
             let mut excludes = self.excludes;
-            let reverse_excludes = excludes
-                .extract_if(|pat| pat.starts_with('!'))
-                .map(|mut pat| {
+            let mut reverse_excludes = vec![];
+            // [TODO] can be simplified and no clone when extract_if stable
+            excludes.retain_mut(|pat| {
+                if pat.starts_with('!') {
                     pat.remove(0);
-                    pat
-                })
-                .collect::<Vec<_>>();
+                    reverse_excludes.push(pat.clone());
+                    false
+                } else {
+                    true
+                }
+            });
             (excludes, reverse_excludes)
         };
 

--- a/fmt/tests/tests.rs
+++ b/fmt/tests/tests.rs
@@ -23,7 +23,7 @@ fn test_remove_file_only_header() {
     let def = defs.get("script_style").unwrap().clone();
     let keywords = vec!["copyright".to_string()];
 
-    let document = parse_header(file, def, &keywords).unwrap();
+    let document = parse_header(file, &def, &keywords).unwrap();
     let end_pos = document.end_pos.unwrap();
     let content = document.file_content.content();
     assert!(content[end_pos..].trim().is_empty());
@@ -36,7 +36,7 @@ fn test_two_headers_should_only_remove_the_first() {
     let def = defs.get("doubleslash_style").unwrap().clone();
     let keywords = vec!["copyright".to_string()];
 
-    let document = parse_header(file, def, &keywords).unwrap();
+    let document = parse_header(file, &def, &keywords).unwrap();
     let end_pos = document.end_pos.unwrap();
     let content = document.file_content.content();
     assert!(content[end_pos..].contains("Copyright 2015 The Prometheus Authors"));

--- a/rust-toolchain.toml
+++ b/rust-toolchain.toml
@@ -13,5 +13,5 @@
 # limitations under the License.
 
 [toolchain]
-channel = "stable"
+channel = "nightly-2024-03-05"
 components = ["cargo", "rustfmt", "clippy", "rust-analyzer"]

--- a/rust-toolchain.toml
+++ b/rust-toolchain.toml
@@ -13,5 +13,5 @@
 # limitations under the License.
 
 [toolchain]
-channel = "nightly-2024-03-05"
+channel = "stable"
 components = ["cargo", "rustfmt", "clippy", "rust-analyzer"]

